File: openc3/lib/openc3/topics/decom_interface_topic.rb
# encoding: ascii-8bit

# Copyright 2023 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

require 'openc3/topics/topic'

module OpenC3
  class DecomInterfaceTopic < Topic
    def self.build_cmd(target_name, cmd_name, cmd_params, range_check, raw, scope:)
      data = {}
      data['target_name'] = target_name.to_s.upcase
      data['cmd_name'] = cmd_name.to_s.upcase
      data['cmd_params'] = cmd_params
      data['range_check'] = range_check
      data['raw'] = raw
      # DecomMicroservice is listening to the DECOMINTERFACE topic and is responsible
      # for actually building the command. This was deliberate to allow this to work
      # with or without an interface.
      decom_id = Topic.write_topic("#{scope}__DECOMINTERFACE__{#{target_name}}",
          { 'build_cmd' => JSON.generate(data, allow_nan: true) }, '*', 100)
      timeout = 5 # Arbitrary 5s timeout
      ack_topic = "{#{scope}__ACKCMD}TARGET__#{target_name}"
      time = Time.now
      while (Time.now - time) < timeout
        Topic.read_topics([ack_topic]) do |topic, msg_id, msg_hash, redis|
          if msg_hash["id"] == decom_id
            if msg_hash["result"] == "SUCCESS"
              return msg_hash
            else
              raise msg_hash["message"]
            end
          end
        end
      end
      raise "Timeout of #{timeout}s waiting for cmd ack. Does target '#{target_name}' exist?"
    end

    def self.inject_tlm(target_name, packet_name, item_hash = nil, type: :CONVERTED, scope:)
      data = {}
      data['target_name'] = target_name.to_s.upcase
      data['packet_name'] = packet_name.to_s.upcase
      data['item_hash'] = item_hash
      data['type'] = type
      Topic.write_topic("#{scope}__DECOMINTERFACE__{#{target_name}}",
          { 'inject_tlm' => JSON.generate(data, allow_nan: true) }, '*', 100)
    end
  end
end

File: openc3/python/openc3/topics/decom_interface_topic.py
# Copyright 2025 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

import json
import time
from openc3.topics.topic import Topic
from openc3.environment import OPENC3_SCOPE
from openc3.utilities.json import JsonEncoder, JsonDecoder


class DecomInterfaceTopic(Topic):
    @classmethod
    def build_cmd(cls, target_name, cmd_name, cmd_params, range_check, raw, timeout=5, scope=OPENC3_SCOPE):
        data = {}
        data["target_name"] = target_name.upper()
        data["cmd_name"] = cmd_name.upper()
        data["cmd_params"] = cmd_params
        data["range_check"] = range_check
        data["raw"] = raw
        # DecomMicroservice is listening to the DECOMINTERFACE topic and is responsible
        # for actually building the command. This was deliberate to allow this to work
        # with or without an interface.
        ack_topic = f"{{{scope}__ACKCMD}}TARGET__{target_name}"
        Topic.update_topic_offsets([ack_topic])
        decom_id = Topic.write_topic(
            f"{scope}__DECOMINTERFACE__{{{target_name}}}",
            {"build_cmd": json.dumps(data, cls=JsonEncoder)},
            "*",
            100,
        )
        start_time = time.time()
        while (time.time() - start_time) < timeout:
            for _topic, _msg_id, msg_hash, _redis in Topic.read_topics([ack_topic]):
                if msg_hash[b"id"] == decom_id:
                    if msg_hash[b"result"] == b"SUCCESS":
                        msg_hash = {k.decode(): v.decode() for (k, v) in msg_hash.items()}
                        msg_hash["buffer"] = json.loads(msg_hash["buffer"], cls=JsonDecoder)
                        return msg_hash
                    else:
                        raise RuntimeError(msg_hash[b"message"])
        raise RuntimeError(f"Timeout of {timeout}s waiting for cmd ack. Does target '{target_name}' exist?")

    @classmethod
    def inject_tlm(
        cls,
        target_name,
        packet_name,
        item_hash=None,
        type="CONVERTED",
        scope=OPENC3_SCOPE,
    ):
        data = {}
        data["target_name"] = target_name.upper()
        data["packet_name"] = packet_name.upper()
        data["item_hash"] = item_hash
        data["type"] = type
        Topic.write_topic(
            f"{scope}__DECOMINTERFACE__{{{target_name}}}",
            {"inject_tlm": json.dumps(data)},
            "*",
            100,
        )
